src/mbgl/style/style_impl.cpp

```diff
@@ -0,0 +1,372 @@
+#include <mbgl/style/style_impl.hpp>
+#include <mbgl/style/observer.hpp>
+#include <mbgl/style/source_impl.hpp>
+#include <mbgl/style/layers/symbol_layer.hpp>
+#include <mbgl/style/layers/custom_layer.hpp>
+#include <mbgl/style/layers/background_layer.hpp>
+#include <mbgl/style/layers/fill_layer.hpp>
+#include <mbgl/style/layers/fill_extrusion_layer.hpp>
+#include <mbgl/style/layers/line_layer.hpp>
+#include <mbgl/style/layers/circle_layer.hpp>
+#include <mbgl/style/layers/raster_layer.hpp>
+#include <mbgl/style/layer_impl.hpp>
+#include <mbgl/style/parser.hpp>
+#include <mbgl/style/transition_options.hpp>
+#include <mbgl/sprite/sprite_loader.hpp>
+#include <mbgl/util/exception.hpp>
+#include <mbgl/util/string.hpp>
+#include <mbgl/util/logging.hpp>
+#include <mbgl/storage/file_source.hpp>
+#include <mbgl/storage/resource.hpp>
+#include <mbgl/storage/response.hpp>
+
+namespace mbgl {
+namespace style {
+
+static Observer nullObserver;
+
+Style::Impl::Impl(Scheduler& scheduler_, FileSource& fileSource_, float pixelRatio)
+    : scheduler(scheduler_),
+      fileSource(fileSource_),
+      spriteLoader(std::make_unique<SpriteLoader>(pixelRatio)),
+      light(std::make_unique<Light>()),
+      observer(&nullObserver) {
+    spriteLoader->setObserver(this);
+    light->setObserver(this);
+}
+
+Style::Impl::~Impl() = default;
+
+void Style::Impl::loadJSON(const std::string& json_) {
+    observer->onStyleLoading();
+
+    url.clear();
+    parse(json_);
+}
+
+void Style::Impl::loadURL(const std::string& url_) {
+    observer->onStyleLoading();
+
+    loaded = false;
+    url = url_;
+
+    styleRequest = nullptr;
```

Comment:
Is this necessary? The next line would already destruct any previous style before assignment right?